### PR TITLE
KAFKA-16921: Migrate all junit 4 code to junit 5 for connect module (part 1)

### PR DESCRIPTION
**Pull request by m1a2st**
https://issues.apache.org/jira/browse/KAFKA-16921

This PR modify connect module

### CODE REVIEW

**Comment by chia7712** [comment]
The junit 4 assertions are not compatible to junit 5, so you ought to fix the conflicts :)

**Comment by chia7712** [comment]
Also, please remove dependency "JUnit Vintage Engine" from connect module

File: build.gradle [hunk 1]
```diff
@@ -3248,7 +3248,6 @@ project(':connect:runtime') {
     testImplementation project(':group-coordinator')
     
     testImplementation libs.junitJupiterApi
-    testImplementation libs.junitVintageEngine
```
Comment (chia7712) [inline build.gradle:3248]: Please cleanup https://github.com/apache/kafka/blob/trunk/build.gradle#L447

**Comment by chia7712** [comment]
Also, please replace `@Category(IntegrationTest.class)` by `@Tag("integrationTest")`  for connect module

File: build.gradle [hunk 2]
```diff
@@ -540,9 +537,6 @@ subprojects {
 
     exclude testsToExclude
 
-    if (shouldUseJUnit5)
-      useJUnitPlatform()
```
Comment (chia7712) [inline build.gradle:540]: Please keep this line ...

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RebalanceSourceConnectorsIntegrationTest.java [hunk 3]
```diff
@@ -350,16 +350,14 @@ private boolean assertConnectorAndTasksAreUniqueAndBalanced() {
...
-            assertNotEquals("Found no tasks running!", maxTasks, 0);
-            assertEquals("Connector assignments are not unique: " + connectors,
-                    connectors.values().size(),
-                    connectors.values().stream().distinct().count());
-            assertEquals("Task assignments are not unique: " + tasks,
-                    tasks.values().size(),
-                    tasks.values().stream().distinct().count());
-            assertTrue("Connectors are imbalanced: " + formatAssignment(connectors), maxConnectors - minConnectors < 2);
-            assertTrue("Tasks are imbalanced: " + formatAssignment(tasks), maxTasks - minTasks < 2);
+            assertNotEquals(maxConnectors, 0, "Found no connectors running!");
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/RebalanceSourceConnectorsIntegrationTest.java]: it should be `assertNotEquals(0, maxConnectors, "Found no connectors running!");`

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RebalanceSourceConnectorsIntegrationTest.java [hunk 4]
```diff
@@ -350,16 +350,14 @@ private boolean assertConnectorAndTasksAreUniqueAndBalanced() {
...
+            assertNotEquals(maxTasks, 0, "Found no tasks running!");
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/RebalanceSourceConnectorsIntegrationTest.java]: ditto

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java [hunk 5]
```diff
@@ -1508,25 +1545,25 @@ private List<Integer> allocations(Function<ConnectorsAndTasks, ? extends Collect
...
-                        new HashSet<>(revocations)
+                        new HashSet<>(revocations),
+                        "Expected no revocations to take place during this round, but connector revocations were issued for worker " + worker
                 )
         );
         returnedAssignments.newlyRevokedTasks().forEach((worker, revocations) ->
                 assertEquals(
-                        "Expected no revocations to take place during this round, but task revocations were issued for worker " + worker,
                         Collections.emptySet(),
-                        new HashSet<>(revocations)
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java]: it can be `assertEquals(0, revocations.size)` or `assertTrue(revocations.isEmpty)`

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java [hunk 6]
```diff
@@ -1508,25 +1545,25 @@ private List<Integer> allocations(Function<ConnectorsAndTasks, ? extends Collect
     private void assertNoRevocations() {
         returnedAssignments.newlyRevokedConnectors().forEach((worker, revocations) ->
                 assertEquals(
-                        "Expected no revocations to take place during this round, but connector revocations were issued for worker " + worker,
                         Collections.emptySet(),
-                        new HashSet<>(revocations)
+                        new HashSet<>(revocations),
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java]: it can be `assertEquals(0, revocations.size)` or `assertTrue(revocations.isEmpty)`

File: build.gradle (hunk 2 above)
Comment (m1a2st) [reply build.gradle:540]: Sorry, I misunderstand the `shouldUseJUnit5` variable logic, so I rollback it

File: build.gradle (hunk 1 above)
Comment (m1a2st) [reply build.gradle:3248]: There is a little problem, If I remove the `testImplementation libs.junitVintageEngine`, Gradle can't get `org.junit.` dependencies and some annotations are error. I find a way to resolve it

Comment (chia7712) [reply build.gradle:3248]: we should remove all usage of junit 4 code form connect module. If you get such error, please rewrite them by junit 5

**Comment by chia7712** [comment]
@m1a2st Could you please split this PR into small PRs (similar to #16253)? It is hard to complete the migration (and review) at once.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java (hunk 5 above)
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java]: We used `assertEquals` here in order to see the difference between expected and actual values if the assertion fails. Can we please keep that and not replace it with assertions about the size or emptiness of the collection instead?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java [hunk 7]
```diff
@@ -1507,26 +1544,24 @@ private List<Integer> allocations(Function<ConnectorsAndTasks, ? extends Collect
 
     private void assertNoRevocations() {
         returnedAssignments.newlyRevokedConnectors().forEach((worker, revocations) ->
-                assertEquals(
-                        "Expected no revocations to take place during this round, but connector revocations were issued for worker " + worker,
-                        Collections.emptySet(),
-                        new HashSet<>(revocations)
+                assertTrue(
+                        revocations.isEmpty(),
+                        "Expected no revocations to take place during this round, but connector revocations were issued for worker " + worker
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java]: IMO we should continue asserting that the collection is equal to an empty set here, since if the assertion fails we get more information that way. This is especially useful when writing new unit tests for this logic and can help save time and headaches during that process.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java [hunk 8]
```diff
@@ -1507,26 +1544,24 @@ private List<Integer> allocations(Function<ConnectorsAndTasks, ? extends Collect
...
                 )
         );
         returnedAssignments.newlyRevokedTasks().forEach((worker, revocations) ->
-                assertEquals(
-                        "Expected no revocations to take place during this round, but task revocations were issued for worker " + worker,
-                        Collections.emptySet(),
-                        new HashSet<>(revocations)
+                assertTrue(
+                        revocations.isEmpty(),
+                        "Expected no revocations to take place during this round, but task revocations were issued for worker " + worker
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java]: Same thought RE asserting that the collection is equal to an empty set.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java (hunk 5 above)
Comment (chia7712) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java]: @C0urante thanks for sharing. That is a good reason so please ignore my previous comments.

> We used assertEquals here in order to see the difference between expected and actual values if the assertion fails.

Just curious. Is it a flaky test? Comparing empty collection is not common assert in code base. I guess there will be a PR to "cleanup" in the future. Hence, maybe we can add more useful message to that assert for debugging

Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java]: These tests aren't flaky, but it is a bit tricky to make changes to them. It definitely saved me time to have these error messages the last time I was in this neck of the woods.

With rudimentary checking logic (searching for `assertEquals(Collections.empty` in the code base), I also see this pattern used 94 times across 16 files in the Connect modules, so it's not exactly uncommon. IMO it'd be better to make it more common in other parts of the code base rather than try to "simplify" things in a way that drops information on assertion failures.

Comment (m1a2st) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java]: @C0urante, thanks for your comments, I will rollback to use `assertEquals()`

Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java]: Thanks!

Comment (chia7712) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java]: > With rudimentary checking logic (searching for assertEquals(Collections.empty in the code base), I also see this pattern used 94 times across 16 files in the Connect modules, so it's not exactly uncommon. IMO it'd be better to make it more common in other parts of the code base rather than try to "simplify" things in a way that drops information on assertion failures.

I file https://issues.apache.org/jira/browse/KAFKA-16929 to have more discussion

File: build.gradle [hunk 9]
```diff
@@ -573,7 +573,7 @@ subprojects {
         useJUnitPlatform {
           includeTags "integration"
           includeTags "org.apache.kafka.test.IntegrationTest"
-	  // Both engines are needed to run JUnit 4 tests alongside JUnit 5 tests.
+          // Both engines are needed to run JUnit 4 tests alongside JUnit 5 tests.
```
Comment (chia7712) [inline build.gradle]: Please revert those unrelated change

File: build.gradle [hunk 10]
```diff
@@ -616,7 +616,7 @@ subprojects {
         useJUnitPlatform {
           excludeTags "integration"
           excludeTags "org.apache.kafka.test.IntegrationTest"
-	  // Both engines are needed to run JUnit 4 tests alongside JUnit 5 tests.
+          // Both engines are needed to run JUnit 4 tests alongside JUnit 5 tests.
```
Comment (chia7712) [inline build.gradle]: Please revert those unrelated change

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/BlockingConnectorTest.java [hunk 11]
```diff
@@ -362,24 +362,16 @@ private void verifyNormalConnector() throws InterruptedException {
     private void assertRequestTimesOut(String requestDescription, ThrowingRunnable request, String expectedTimeoutMessage) {
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/BlockingConnectorTest.java]: `ThrowingRunnable` should be replaced by `Executable`

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/StateTrackerTest.java [hunk 12]
```diff
@@ -100,4 +100,4 @@ public void calculateDurations() {
 
     }
 
-}
\ No newline at end of file
+}
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/StateTrackerTest.java]: Please revert this unrelated change

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java [hunk 13]
```diff
@@ -1889,4 +1888,4 @@ private void assertTaskMetricValue(String name, String expected) {
         String measured = metrics.currentMetricValueAsString(taskGroup, name);
         assertEquals(expected, measured);
     }
-}
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java]: Please revert this unrelated change

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/LoggingResourceTest.java [hunk 14]
```diff
@@ -162,4 +162,4 @@ public void testSetLevelClusterScope() {
         verify(herder).setClusterLoggerLevel(logger, level);
     }
 
-}
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/LoggingResourceTest.java]: Please revert this unrelated change

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for updated PR. It seems those changed files are still using `@RunWith(MockitoJUnitRunner.StrictStubs.class)`. Could you please update them?

**Comment by m1a2st** [comment]
@chia7712, Thanks for your comment, There are two classes still have `@RunWith(MockitoJUnitRunner.StrictStubs.class)` one is `LoggingResourceTest` and `RestClientTest$Test`, but these two classes modify to 
```
@ExtendWith(MockitoExtension.class)
@MockitoSettings(strictness = Strictness.STRICT_STUBS)
```
gradle build test file will error, like 
```
No matching tests found in any candidate test task.
    Requested tests:
        Test pattern org.apache.kafka.connect.runtime.rest.RestClientTest$RequestFailureParameterizedTest in task :connect:runtime:test
 ```
I think that it need to modify some gradle task to resolve it.

**Comment by chia7712** [comment]
> I think that it need to modify some gradle task to resolve it.

yep, `connect:runtime` can use junit 5 now. Sorry that my previous comment is not clear. Please complete following changes:

1. Please remove `shouldUseJUnit5` from build.gradle.
2. please make sure `connect:runtime` has following dependencies:
```
    testImplementation libs.junitJupiter
    testImplementation libs.junitVintageEngine
    testImplementation libs.mockitoJunitJupiter
    testImplementation libs.mockitoCore
```
3. `junitJupiterApi` is included by `libs.junitJupiter` so we can remove it.

**Comment by m1a2st** [comment]
@chia7712, Thanks for your comment, I remove all `@RunWith(MockitoJUnitRunner.StrictStubs.class)`, PTAL

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java [hunk 15]
```diff
@@ -1228,7 +1228,7 @@ private void testConfigProviderRegex(String rawConnConfig) {
     private void testConfigProviderRegex(String rawConnConfig, boolean expected) {
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java]: this test `AbstractHerderTest` is still using junit 4.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractWorkerSourceTaskTest.java [hunk 16]
```diff
@@ -90,12 +90,12 @@
 import static org.apache.kafka.connect.runtime.TopicCreationConfig.PARTITIONS_CONFIG;
 import static org.apache.kafka.connect.runtime.TopicCreationConfig.REPLICATION_FACTOR_CONFIG;
 import static org.apache.kafka.connect.runtime.WorkerConfig.TOPIC_CREATION_ENABLE_CONFIG;
-import static org.junit.Assert.assertArrayEquals;
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractWorkerSourceTaskTest.java]: this test `AbstractWorkerSourceTaskTest` is still using junit 4.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/BlockingConnectorTest.java [hunk 17]
```diff
@@ -43,7 +43,7 @@
 import org.junit.Before;
 import org.junit.Test;
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/BlockingConnectorTest.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for updated patch. I don't review all files, but it seems many files are still using junit 4 code. Could you please check them again?

**Review by chia7712 — APPROVED** [review]
LGTM

@m1a2st thanks for reducing the size of PR.
